xarray/backends/zarr.py

```diff
@@ -0,0 +1,554 @@
+                 writer=None, autoclose=None):
+        self._mode = mode
+        self._synchronizer = synchronizer
+        self._group = group
+
+        # zarr stores don't need to be opened, closed, or synced.
+        # So what do we do with all this logical about openers?
+        if autoclose:
+            raise NotImplementedError('autoclose=True is not implemented '
+                                      'for the zarr backend')
+        self._autoclose = False
+        self._isopen = True
+
+        opener = functools.partial(_open_zarr_group, store=store,
+                                   synchronizer=synchronizer, group=group)
+        self._opener = opener
+        self.ds = self._opener(mode=mode)
+
+        # initialize hidden dimension attribute
+        if _DIMENSION_KEY not in self.ds.attrs:
+            self.ds.attrs[_DIMENSION_KEY] = {}
+
+        # do we need to define attributes for all of the opener keyword args?
+        super(ZarrStore, self).__init__(writer)
+
+    def open_store_variable(self, name, zarr_array):
+        # I don't see why it is necessary to wrap self.ds[name]
+        # zarr seems to implement the required ndarray interface
+        # TODO: possibly wrap zarr array in dask with aligned chunks
+        data = indexing.LazilyIndexedArray(ZarrArrayWrapper(name, self))
+        dimensions, attributes = _get_zarr_dims_and_attrs(zarr_array,
+                                                          _DIMENSION_KEY)
+        attributes = _decode_zarr_attrs(attributes)
+        encoding = {'chunks': zarr_array.chunks,
+                    'compressor': zarr_array.compressor,
+                    'filters': zarr_array.filters}
+        # _FillValue needs to be in attributes, not encoding, so it will get
+        # picked up by decode_cf
+        if getattr(zarr_array, 'fill_value') is not None:
+            attributes['_FillValue'] = zarr_array.fill_value
+
+        return Variable(dimensions, data, attributes, encoding)
+
+    def get_variables(self):
+        with self.ensure_open(autoclose=False):
+            return FrozenOrderedDict((k, self.open_store_variable(k, v))
+                                     for k, v in self.ds.arrays())
+
+    def get_attrs(self):
+        with self.ensure_open(autoclose=True):
+            _, attributes = _get_zarr_dims_and_attrs(self.ds, _DIMENSION_KEY)
+            return _decode_zarr_attrs(attributes)
+
+    def get_dimensions(self):
+        with self.ensure_open(autoclose=True):
+            dimensions, _ = _get_zarr_dims_and_attrs(self.ds, _DIMENSION_KEY)
+            return dimensions
+
+    def set_dimension(self, name, length, is_unlimited=False):
+        if is_unlimited:
+            raise NotImplementedError(
+                "Zarr backend doesn't know how to handle unlimited dimensions")
+        with self.ensure_open(autoclose=False):
+            self.ds.attrs[_DIMENSION_KEY][name] = length
```

Comment:
This probably needs to do a consistency check first to ensure dimensions aren't being changed. That could happen if you write to the same zarr file multiple times.